### ISSUE TITLE
Add TribitsCTestDriverCore.cmake support to TribitsExampleProject and documenation and testing

### ISSUE THREAD

**Issue by bartlettroscoe** [TriBITSPub/TriBITS#193]
**Blocks:** #183

**Description:**

This story will be to add support for TribitsCTestDriverCore.cmake for CTEst -S scripts to submit to CDash for the TribitsExampleProject.  Along with this, a HowTo for adding this support will be added to the TriBITS Developers Guide.  Also, this will provide a way to add very strong testing for the TribitsCTestDriveCore.cmake file (everything except the CDash submit in automated testing).

**Comment by bartlettroscoe** [comment]
I pushed a merged in topic branch that has completed this as of commit c8c9a81.

I will now put this in review for a bit to see how things shake out.  This will not be fully tested until merged with Trilinos automated testing.  I might need to be careful about doing some testing with Trilinos before closing this issue.

----

**DETAILED NOTES:**

I need to get this done ASAP so that I can refactor TribitsCTestDriverCore.cmake to do a single configure, build, test, and submit to CDash in #183.

**1) Add support for TribitsCTestDriverCore.cmake to TribitsExampleProject and test with local `make dashboard`**

Wow, to get `make dashboard` working, I just had to create the file TribitsExampleProject/CTestConfig.cmake:

```
INCLUDE(SetDefaultAndFromEnv)

SET(CTEST_NIGHTLY_START_TIME "04:00:00 UTC") # 10 PM MDT or 9 PM MST

IF (NOT DEFINED CTEST_DROP_METHOD)
  SET_DEFAULT_AND_FROM_ENV(CTEST_DROP_METHOD "http")
ENDIF()

IF (CTEST_DROP_METHOD STREQUAL "http")
  SET_DEFAULT_AND_FROM_ENV(CTEST_DROP_SITE "my.cdash.org")
  SET_DEFAULT_AND_FROM_ENV(CTEST_PROJECT_NAME "TribitsExampleProject")
  SET_DEFAULT_AND_FROM_ENV(CTEST_DROP_LOCATION "/submit.php?project=TribitsExampleProject")
  SET_DEFAULT_AND_FROM_ENV(CTEST_TRIGGER_SITE "")
  SET_DEFAULT_AND_FROM_ENV(CTEST_DROP_SITE_CDASH TRUE)
ENDIF()
```

Actually, that file was already there.  It was just submitting to a "TribitsExProj" CDash project on the casl-dev CDash site.

[Done]

**2) Create TribitsExampleProject CDash project on my.cdash.org and do some trial submits:**

I create the CDash project:

* http://my.cdash.org/index.php?project=TribitsExampleProject

That only took about 5 mintues!  I was able to do a `make dashboard` submit to it right away!  You can see that submission at:

* http://my.cdash.org/index.php?project=TribitsExampleProject&parentid=1212298

That was easy!

I was also able to submit to the existing CDash project `TribitsExProj` on the Trilinos testing site by configuring TribitsExampleProject with:

```
-DCTEST_DROP_SITE=testing.sandia.gov/cdash \
-DCTEST_PROJECT_NAME=TribitsExProj \
-DCTEST_DROP_LOCATION="/submit.php?project=TribitsExProj" \
```

and then ``make dashboard``.  You can see the results of that, for example, here:

* http://testing.sandia.gov/cdash/index.php?project=TribitsExProj&parentid=2914030

[Done]

**3) Add CTest -S driver script for custom submit script:**

We also need an example of a custom CTest -S driver script that would be run by a cron job or by a Jenkins job.  I think I would like the Travis CI driver for TriBITS to run an automated invocation of a few ctest -S driver scripts for TribitsExampleProject and submit to my.cdash.org.  That could be a nice example and it would populate that CDash project to provide examples for users of TriBITS.

Since the TribitsCTestDriverCore.cmake code does not dictate how CTest -S scripts are set up, I will demonstrate a simple system that avoids duplication.

Here is my first trial:

```
$ mkdir MOCK_TRIBITSEXPROJ_SERIAL_DEBUG
$ cd MOCK_TRIBITSEXPROJ_SERIAL_DEBUG/
$ export TRIBITS_DIR=../../../Trilinos/TriBITS/tribits
$ ln -s $TRIBITS_DIR/examples/TribitsExampleProject .
$ time env  CTEST_DASHBOARD_ROOT=$PWD \
     CTEST_DROP_SITE=testing.sandia.gov/cdash \
     CTEST_PROJECT_NAME=TribitsExProj \
     CTEST_DROP_LOCATION="/submit.php?project=TribitsExProj" \
     CTEST_TEST_TYPE=Experimental \
     CTEST_DO_SUBMIT=ON \
     CTEST_DO_UPDATES=OFF \
     CTEST_START_WITH_EMPTY_BINARY_DIRECTORY=TRUE \
     ctest -V -S \
      $TRIBITS_DIR/examples/TribitsExampleProject/cmake/ctest/general_gcc/ctest_serial_debug.cmake \
     &> console.out

real    0m24.694s
user    0m3.256s
sys     0m1.469s

```

and that submitted to:

* http://testing.sandia.gov/cdash/index.php?project=TribitsExProj&parentid=2914104

Wow, that was not very hard either!

**4) Add HowTo to TribitsDevelopersGuide on how to add support for CTest -S driver scripts to submit to CDash.**

[Done]
 
**5) Add some automated tests for TribitsCTestDriverCore.cmake with TribitsExampleProject under TriBITS/tests/ctest_driver/:**

Now I just need to add a few automated tests to pin down more the full behavior of TribitsCTestDriverCore.cmake before a jump into the major refactoring that will be a part of 

ToDo:

* Add option TriBITS_TRIBITEXPROJ_SUBMIT_TO with values "", "MY_CDASH" and "TRILINOS_CDASH" to turn on the submit of one of the TribitsExampleProject CTest -S script submits to an actual CDash site. ... That works fine locally but just hung when run by Travis CI [Done]

* Add ALWAYS_FAIL_ON_ZERO_RETURN to TRIBITS_ADD_ADVANCED_TEST() to require a non-zero return code as a critera for passing the tests in addition to other criteria [Done]

* Turn of FATAL_ERROR and debug why failing CTest Driver runs tests don't fail!

* Strengthen greps in tests [Done]

* Fix failing tests for CMake 2.8.11 [Done]

I have configured this with:

```
$ cat do-configure
#!/bin/bash

../../../Trilinos/TriBITS/dev_testing/generic/do-configure-mpi-debug \
-DCTEST_PARALLEL_LEVEL=16 \
-DTriBITS_ENABLE_CONFIGURE_TIMING=ON \
"$@"

$ ./do-configure -DTriBITS_TRIBITSEXPROJ_SUBMIT_TO=TRILINOS_CDASH
```

with CMake 3.6.2 and submitted the TribitsCTestDriverCore.cmake testing cases to CDash that displayed here:

* http://testing.sandia.gov/cdash/index.php?project=TribitsExProj&parentid=2924891&filtercount=2&showfilters=1&field1=buildstarttime&compare1=84&value1=now&filtercombine=and

I then cleaned up the committs with `git rebase -i github/master` and pushed with the checkin-test-crf450.sh script.

**Comment by bartlettroscoe** [comment]
This has been done for a long time.  There are automated tests protecting this and people have been using the documentation.  Since I don't think I am going to get any formal reviews on this, I am going to close.
